# Data Preparation

In [1]:
#import related libraries
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler

#loading data
file_name = "https://raw.githubusercontent.com/eozhisarcikli/model-of-customer-churn/main/data/WA_Fn-UseC_-Telco-Customer-Churn.csv"
churn_df = pd.read_csv(file_name) 

#quick look to data
churn_df.head()

customerID  gender  SeniorCitizen  ... MonthlyCharges TotalCharges  Churn
0  7590-VHVEG  Female              0  ...          29.85        29.85     No
1  5575-GNVDE    Male              0  ...          56.95       1889.5     No
2  3668-QPYBK    Male              0  ...          53.85       108.15    Yes
3  7795-CFOCW    Male              0  ...          42.30      1840.75     No
4  9237-HQITU  Female              0  ...          70.70       151.65    Yes

[5 rows x 21 columns]

In [ ]:
#check for empty fields(Note; " " is not Null but a spaced character)
len(churn_df[churn_df['TotalCharges'] == " "])

11

In [ ]:
#drop missing data
churn_df = churn_df[churn_df['TotalCharges'] != " "]
#recheck
len(churn_df[churn_df['TotalCharges'] == " "])

0

In [ ]:
#customer id col
Id_col = ['customerID']

#Target columns
target_col = ["Churn"]

#categorical columns
cat_cols = churn_df.nunique()[churn_df.nunique() < 6].keys().tolist()
cat_cols = [x for x in cat_cols if x not in target_col + Id_col]

#numerical columns
num_cols = [x for x in churn_df.columns if x not in cat_cols + target_col + Id_col]

#Binary columns with 2 values
bin_cols = churn_df.nunique()[churn_df.nunique() == 2].keys().tolist()

#Columns more than 2 values
multi_cols = [i for i in cat_cols if i not in bin_cols]

#Label encoding Binary columns
le = LabelEncoder()
for i in bin_cols :
    churn_df[i] = le.fit_transform(churn_df[i])
    
#Duplicating columns for multi value columns
churn_df = pd.get_dummies(data = churn_df, columns = multi_cols )
churn_df.head()

customerID  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  7590-VHVEG       0              0        1           0       1   
1  5575-GNVDE       1              0        0           0      34   
2  3668-QPYBK       1              0        0           0       2   
3  7795-CFOCW       1              0        0           0      45   
4  9237-HQITU       0              0        0           0       2   

   PhoneService  PaperlessBilling  MonthlyCharges TotalCharges  ...  \
0             0                 1           29.85        29.85  ...   
1             1                 0           56.95       1889.5  ...   
2             1                 1           53.85       108.15  ...   
3             0                 0           42.30      1840.75  ...   
4             1                 1           70.70       151.65  ...   

   StreamingMovies_No  StreamingMovies_No internet service  \
0                   1                                    0   
1                   1                                    0   
2                   1                                    0   
3                   1                                    0   
4                   1                                    0   

   StreamingMovies_Yes  Contract_Month-to-month  Contract_One year  \
0                    0                        1                  0   
1                    0                        0                  1   
2                    0                        1                  0   
3                    0                        0                  1   
4                    0                        1                  0   

   Contract_Two year  PaymentMethod_Bank transfer (automatic)  \
0                  0                                        0   
1                  0                                        0   
2                  0                                        0   
3                  0                                        1   
4                  0                                        0   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                      0                               1   
1                                      0                               0   
2                                      0                               0   
3                                      0                               0   
4                                      0                               1   

   PaymentMethod_Mailed check  
0                           0  
1                           1  
2                           1  
3                           0  
4                           0  

[5 rows x 42 columns]

In [ ]:
#Scaling Numerical columns
std = StandardScaler()

# Scale data
scaled = std.fit_transform(churn_df[num_cols])
scaled = pd.DataFrame(scaled,columns=num_cols)

#dropping original values merging scaled values for numerical columns
df_telcom_og = churn_df.copy()
churn_df = churn_df.drop(columns = num_cols,axis = 1)
churn_df = churn_df.merge(scaled, left_index=True, right_index=True, how = "left")

#churn_df.info()
churn_df.head()

customerID  gender  SeniorCitizen  Partner  Dependents  PhoneService  \
0  7590-VHVEG       0              0        1           0             0   
1  5575-GNVDE       1              0        0           0             1   
2  3668-QPYBK       1              0        0           0             1   
3  7795-CFOCW       1              0        0           0             0   
4  9237-HQITU       0              0        0           0             1   

   PaperlessBilling  Churn  MultipleLines_No  MultipleLines_No phone service  \
0                 1      0                 0                               1   
1                 0      0                 1                               0   
2                 1      1                 1                               0   
3                 0      0                 0                               1   
4                 1      1                 1                               0   

   ...  Contract_Month-to-month  Contract_One year  Contract_Two year  \
0  ...                        1                  0                  0   
1  ...                        0                  1                  0   
2  ...                        1                  0                  0   
3  ...                        0                  1                  0   
4  ...                        1                  0                  0   

   PaymentMethod_Bank transfer (automatic)  \
0                                        0   
1                                        0   
2                                        0   
3                                        1   
4                                        0   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                      0                               1   
1                                      0                               0   
2                                      0                               0   
3                                      0                               0   
4                                      0                               1   

   PaymentMethod_Mailed check    tenure  MonthlyCharges  TotalCharges  
0                           0 -1.280248       -1.161694     -0.994194  
1                           1  0.064303       -0.260878     -0.173740  
2                           1 -1.239504       -0.363923     -0.959649  
3                           0  0.512486       -0.747850     -0.195248  
4                           0 -1.239504        0.196178     -0.940457  

[5 rows x 42 columns]

In [ ]:
churn_df.drop(['customerID'], axis=1, inplace=True) #delete customerID, which is irrelevant for modelling
churn_df.head()

gender  SeniorCitizen  Partner  Dependents  PhoneService  PaperlessBilling  \
0       0              0        1           0             0                 1   
1       1              0        0           0             1                 0   
2       1              0        0           0             1                 1   
3       1              0        0           0             0                 0   
4       0              0        0           0             1                 1   

   Churn  MultipleLines_No  MultipleLines_No phone service  MultipleLines_Yes  \
0      0                 0                               1                  0   
1      0                 1                               0                  0   
2      1                 1                               0                  0   
3      0                 0                               1                  0   
4      1                 1                               0                  0   

   ...  Contract_Month-to-month  Contract_One year  Contract_Two year  \
0  ...                        1                  0                  0   
1  ...                        0                  1                  0   
2  ...                        1                  0                  0   
3  ...                        0                  1                  0   
4  ...                        1                  0                  0   

   PaymentMethod_Bank transfer (automatic)  \
0                                        0   
1                                        0   
2                                        0   
3                                        1   
4                                        0   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                      0                               1   
1                                      0                               0   
2                                      0                               0   
3                                      0                               0   
4                                      0                               1   

   PaymentMethod_Mailed check    tenure  MonthlyCharges  TotalCharges  
0                           0 -1.280248       -1.161694     -0.994194  
1                           1  0.064303       -0.260878     -0.173740  
2                           1 -1.239504       -0.363923     -0.959649  
3                           0  0.512486       -0.747850     -0.195248  
4                           0 -1.239504        0.196178     -0.940457  

[5 rows x 41 columns]

In [ ]:
churn_df[churn_df.isnull().any(axis=1)] #last time check for null values

gender  SeniorCitizen  Partner  Dependents  PhoneService  \
7032       1              1        0           0             1   
7033       1              0        0           0             1   
7034       0              0        0           0             1   
7035       1              0        0           0             1   
7036       0              0        0           0             0   
7037       0              0        0           0             1   
7038       1              0        1           1             1   
7039       0              0        1           1             1   
7040       0              0        1           1             0   
7041       1              1        1           0             1   
7042       1              0        0           0             1   

      PaperlessBilling  Churn  MultipleLines_No  \
7032                 1      1                 0   
7033                 1      0                 1   
7034                 1      1                 0   
7035                 1      0                 1   
7036                 0      0                 0   
7037                 1      0                 1   
7038                 1      0                 0   
7039                 1      0                 0   
7040                 1      0                 0   
7041                 1      1                 0   
7042                 1      0                 1   

      MultipleLines_No phone service  MultipleLines_Yes  ...  \
7032                               0                  1  ...   
7033                               0                  0  ...   
7034                               0                  1  ...   
7035                               0                  0  ...   
7036                               1                  0  ...   
7037                               0                  0  ...   
7038                               0                  1  ...   
7039                               0                  1  ...   
7040                               1                  0  ...   
7041                               0                  1  ...   
7042                               0                  0  ...   

      Contract_Month-to-month  Contract_One year  Contract_Two year  \
7032                        1                  0                  0   
7033                        1                  0                  0   
7034                        1                  0                  0   
7035                        1                  0                  0   
7036                        0                  1                  0   
7037                        0                  0                  1   
7038                        0                  1                  0   
7039                        0                  1                  0   
7040                        1                  0                  0   
7041                        1                  0                  0   
7042                        0                  0                  1   

      PaymentMethod_Bank transfer (automatic)  \
7032                                        0   
7033                                        0   
7034                                        0   
7035                                        1   
7036                                        0   
7037                                        1   
7038                                        0   
7039                                        0   
7040                                        0   
7041                                        0   
7042                                        1   

      PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
7032                                      0                               1   
7033                                      1                               0   
7034                                      1                               0   
7035                                      0                              

In [ ]:
churn_df = churn_df.dropna() #delete null values because only eleven values
churn_df[churn_df.isnull().any(axis=1)] #double check that nulls either removed or not

Empty DataFrame
Columns: [gender, SeniorCitizen, Partner, Dependents, PhoneService, PaperlessBilling, Churn, MultipleLines_No, MultipleLines_No phone service, MultipleLines_Yes, InternetService_DSL, InternetService_Fiber optic, InternetService_No, OnlineSecurity_No, OnlineSecurity_No internet service, OnlineSecurity_Yes, OnlineBackup_No, OnlineBackup_No internet service, OnlineBackup_Yes, DeviceProtection_No, DeviceProtection_No internet service, DeviceProtection_Yes, TechSupport_No, TechSupport_No internet service, TechSupport_Yes, StreamingTV_No, StreamingTV_No internet service, StreamingTV_Yes, StreamingMovies_No, StreamingMovies_No internet service, StreamingMovies_Yes, Contract_Month-to-month, Contract_One year, Contract_Two year, PaymentMethod_Bank transfer (automatic), PaymentMethod_Credit card (automatic), PaymentMethod_Electronic check, PaymentMethod_Mailed check, tenure, MonthlyCharges, TotalCharges]
Index: []

[0 rows x 41 columns]

# Modelling (Train/Test)

In [ ]:
from sklearn.model_selection import train_test_split

#remove label values from training data
X = churn_df.drop(['Churn'], axis=1).values

#assign those label values to Y dataset
y = churn_df['Churn'].values

#split it to a 70(train):30(test) ratio
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [ ]:
df_train = pd.DataFrame(X_train)
df_train.head()

0    1    2    3    4    5    6    7    8    9   ...   30   31   32   33  \
0  1.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  ...  1.0  0.0  0.0  0.0   
1  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  1.0  ...  1.0  0.0  0.0  0.0   
2  1.0  1.0  1.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0  ...  1.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0  ...  1.0  0.0  0.0  0.0   
4  0.0  0.0  1.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0  ...  0.0  1.0  0.0  0.0   

    34   35   36        37        38        39  
0  1.0  0.0  0.0  1.571829  0.417227  1.441271  
1  0.0  0.0  1.0 -0.791321 -1.475816 -0.896339  
2  0.0  1.0  0.0 -0.546857  1.170123 -0.174115  
3  0.0  1.0  0.0  1.571829  0.944088  1.863068  
4  1.0  0.0  0.0  1.408853  0.028314  0.899824  

[5 rows x 40 columns]

# Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

model = LogisticRegression()
model.fit(X_train, y_train)

predictions = model.predict(X_test)
score = model.score(X_test, y_test)

print("Accuracy = " + str(score))
print(classification_report(y_test, predictions))

Accuracy = 0.791172282866635
              precision    recall  f1-score   support

           0       0.83      0.90      0.86      1551
           1       0.63      0.49      0.55       556

    accuracy                           0.79      2107
   macro avg       0.73      0.70      0.71      2107
weighted avg       0.78      0.79      0.78      2107



So accuracy shows 79% right now. Not too bad, actually. But recall, which is precision in predicting where the customer has churned, isn't too great. Recall measures, how much positive samples of our dataset did we identify correct. And recall[1] is actually at 49%, which is not that great. It is easy to get recall[0] (90%), because there are lot more data in this sample. Unfortunately we are not trying to get this one. So while that classifier is doing a fairly good job, it's not ideal.

# Random Forests

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

model_rf = RandomForestClassifier()
model_rf.fit(X_train, y_train)

predictions = model_rf.predict(X_test)
score = model_rf.score(X_test, y_test)

print("Accuracy = " + str(score))
print(classification_report(y_test, predictions))

Accuracy = 0.7726625533934504
[[1391  160]
 [ 319  237]]
              precision    recall  f1-score   support

           0       0.81      0.90      0.85      1551
           1       0.60      0.43      0.50       556

    accuracy                           0.77      2107
   macro avg       0.71      0.66      0.68      2107
weighted avg       0.76      0.77      0.76      2107



Random forest also didn't do a good job. Accuracy actually went down to 77%. And recall[1] is even worse with 43%.